lib/galaxy/tools/parameters/__init__.py

```diff
@@ -312,7 +312,70 @@ def populate_state(request_context, inputs, incoming, state, errors={}, prefix='
     3
     >>> print(state['b'][0]['d'][0]['f']['h'])
     4
+    >>> # now test with input_format='21.01'
+    >>> nested = {'a': 1, 'b': [{'c': 2, 'd': [{'e': 3, 'f': {'h': 4, 'g': True}}]}]}
+    >>> state_new = OrderedDict()
+    >>> populate_state(trans, inputs, nested, state_new, check=False, input_format='21.01')
+    >>> print(state_new['a'])
+    1
+    >>> print(state_new['b'][0]['c'])
+    2
+    >>> print(state_new['b'][0]['d'][0]['e'])
+    3
+    >>> print(state_new['b'][0]['d'][0]['f']['h'])
+    4
+
     """
+    if input_format == 'legacy':
+        _populate_state_legacy(request_context, inputs, incoming, state, errors=errors, context=context, check=check, simple_errors=simple_errors)
+        return
+    elif input_format == '21.01':
+        context = ExpressionContext(state, context)
+        for input in inputs.values():
+            state[input.name] = input.get_initial_value(request_context, context)
+            group_state = state[input.name]
+            if input.type == 'repeat':
+                if len(incoming[input.name]) > input.max or len(incoming[input.name]) < input.min:
+                    errors[input.name] = 'The number of repeat elements is outside the range specified by the tool.'
+                else:
+                    del group_state[:]
+                    for rep_index, rep in enumerate(incoming[input.name]):
+                        new_state = {}
+                        group_state.append(new_state)
+                        populate_state(request_context, input.inputs, rep, new_state, errors, context=context, check=check, simple_errors=simple_errors, input_format='21.01')
+
+            elif input.type == 'conditional':
+                test_param_value = incoming.get(input.name).get(input.test_param.name)
+                value, error = check_param(request_context, input.test_param, test_param_value, context, simple_errors=simple_errors) if check else [test_param_value, None]
+                if error:
+                    errors[input.test_param.name] = error
```

Comment:
Now that I look at this again, maybe it would be good to keep passing around the `prefix` to have more precise keys in the `errors` dict?

Reply:
I don't think that's necessary, if I understand this correctly we should be getting back a nested dict with errors here, that is much simpler to translate back in a frontend model for instance, or to relate to your input parameters.

Reply:
@mvdbeek I don't think the `errors` dict is nested in the current implementation.

Reply:
Maybe something like this? https://github.com/galaxyproject/galaxy/pull/10720/commits/5f00cf6ae4246e48c278f88b3e4770b374b2182d